compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -2193,46 +2195,76 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             && let hir::FnRetTy::Return(ty) = decl.output
             && let hir::TyKind::InferDelegation(_, _) = ty.kind
         {
-            try_emit("recursive delegation");
+            try_emit("recursive delegation is not supported yet");
         }
 
         let sig = self.tcx().fn_sig(sig_id).instantiate_identity();
         if sig.output().has_opaque_types() {
-            try_emit("delegation to a function with opaque type");
+            try_emit("delegation to a function with opaque type is not supported yet");
         }
 
         let sig_generics = self.tcx().generics_of(sig_id);
         let parent = self.tcx().parent(self.item_def_id());
         let parent_generics = self.tcx().generics_of(parent);
 
-        let parent_is_trait = (self.tcx().def_kind(parent) == DefKind::Trait) as usize;
+        let parent_def_kind = self.tcx().def_kind(parent);
+        let parent_is_trait = (parent_def_kind == DefKind::Trait) as usize;
         let sig_has_self = sig_generics.has_self as usize;
 
         if sig_generics.count() > sig_has_self || parent_generics.count() > parent_is_trait {
-            try_emit("delegation with early bound generics");
+            try_emit("delegation with early bound generics is not supported yet");
+        }
+
+        // There is no way to instantiate `Self` param for caller if
+        // 1. callee is a trait method && callee contains `Self` param
+        // 2. parent item isn't a trait or impl
+        if let DefKind::AssocFn = self.tcx().def_kind(sig_id)
+            && !matches!(parent_def_kind, DefKind::Impl { .. } | DefKind::Trait)
+            && self.tcx().associated_item(sig_id).container
+                == ty::AssocItemContainer::TraitContainer
+        {
+            struct HasSelfParamVisitor;
+
+            impl<'tcx> ty::TypeVisitor<TyCtxt<'tcx>> for HasSelfParamVisitor {
```

Comment:
Do you plan to eventually produce a different signature for `reuse Trait::foo` depending on the use of `Self` in signature?

That doesn't seem right.
The number of generic parameters should probably be the same regardless of specific input/output types.

Reply:
`Self` parameter from `Trait::foo` will turn either into 1 or into 0 parameters, depending on the details of
https://github.com/rust-lang/rfcs/blob/923b8e65639de80a93ef2b9770ba1c976c24cf5d/text/0000-fn-delegation.md#self-mapping

Reply:
> Do you plan to eventually produce a different signature for `reuse Trait::foo` depending on the use of `Self` in signature?
> 
> That doesn't seem right. The number of generic parameters should probably be the same regardless of specific input/output types.

At the moment, I have banned all cases where a free function delegates to a trait method. If we plan to support this case in the future, then I expect for delegation without `qself` to map `Self` to `T: Trait` like an example from description. For callee signature without `Self` new parameter will be unused.
```rust
trait Trait {
    fn bar(x: i32) {...}
}

reuse Trait::foo;
// will be desugared to:
fn bar<T: Trait>(x: i32) { Trait::foo(x) }
```

That is why this case is working on master and this check has been added.